File: lib/code-parser.js
#!/usr/bin/env node

/**
 * Code Flow Parser for openstack-k8s-operators Operators
 * Analyzes Go operator code to extract flow patterns
 */

const fs = require('fs');
const path = require('path');

class OperatorFlowParser {
    constructor() {
        this.controllers = [];
        this.reconcileFunctions = [];
        this.customResources = [];
        this.webhooks = [];
    }

    // Parse operator directory structure
    parseOperatorDirectory(dir) {
        try {
            const result = {
                controllers: this.findControllers(dir),
                reconcilers: this.findReconcilers(dir),
                crds: this.findCRDs(dir),
                webhooks: this.findWebhooks(dir),
                main: this.findMainFunction(dir)
            };
            
            console.log(JSON.stringify(result, null, 2));
            return result;
        } catch (error) {
            console.error('Error parsing operator:', error.message);
            return null;
        }
    }

    // Find controller files
    findControllers(dir) {
        const controllers = [];
        const controllerPattern = /controller|reconciler/i;
        
        try {
            this.walkDirectory(dir, (filePath) => {
                if (filePath.endsWith('.go') && controllerPattern.test(filePath)) {
                    const content = fs.readFileSync(filePath, 'utf8');
                    const controller = this.parseControllerFile(filePath, content);
                    if (controller) {
                        controllers.push(controller);
                    }
                }
            });
        } catch (error) {
            console.error('Error finding controllers:', error.message);
        }

        return controllers;
    }

    // Parse individual controller file
    parseControllerFile(filePath, content) {
        const reconcilePattern = /func\s+\(.*?\)\s+Reconcile\s*\([^)]*\)\s+\([^)]*\)/g;
        const setupPattern = /func\s+\([^)]+\)\s+SetupWithManager\s*\([^)]*\)/g;
        
        const reconcileFunctions = [];
        const setupFunctions = [];
        
        let match;
        
        // Find Reconcile functions
        while ((match = reconcilePattern.exec(content)) !== null) {
            reconcileFunctions.push({
                signature: match[0],
                line: this.getLineNumber(content, match.index)
            });
        }
        
        // Find SetupWithManager functions
        while ((match = setupPattern.exec(content)) !== null) {
            setupFunctions.push({
                signature: match[0],
                line: this.getLineNumber(content, match.index)
            });
        }
        
        if (reconcileFunctions.length > 0 || setupFunctions.length > 0) {
            return {
                file: path.relative(process.cwd(), filePath),
                reconcile: reconcileFunctions,
                setup: setupFunctions,
                imports: this.extractImports(content),
                structs: this.extractStructs(content)
            };
        }
        
        return null;
    }

    // Find reconciler functions and their flow
    findReconcilers(dir) {
        const reconcilers = [];
        
        try {
            this.walkDirectory(dir, (filePath) => {
                if (filePath.endsWith('.go')) {
                    const content = fs.readFileSync(filePath, 'utf8');
                    const flows = this.parseReconcileFlow(filePath, content);
                    if (flows.length > 0) {
                        reconcilers.push({
                            file: path.relative(process.cwd(), filePath),
                            flows: flows
                        });
                    }
                }
            });
        } catch (error) {
            console.error('Error finding reconcilers:', error.message);
        }

        return reconcilers;
    }

    // Parse reconcile function flow
    parseReconcileFlow(filePath, content) {
        const flows = [];
        const signaturePattern = /func\s+\([^)]+\)\s+Reconcile\s*\([^)]*\)\s+\([^)]*\)\s*{/g;

        let match;
        while ((match = signaturePattern.exec(content)) !== null) {
            const bodyStart = match.index + match[0].length;
            const body = this.extractFunctionBody(content, bodyStart);
            const flow = {
                function: match[0].trimEnd().slice(0, -1).trim() + ' {...}',
                line: this.getLineNumber(content, match.index),
                steps: this.extractFlowSteps(body),
                errorHandling: this.extractErrorHandling(body),
                returns: this.extractReturns(body)
            };
            flows.push(flow);
        }

        return flows;
    }

    // Extract function body using brace counting
    extractFunctionBody(content, startIndex) {
        let depth = 1;
        let i = startIndex;
        while (i < content.length && depth > 0) {
            if (content[i] === '{') depth++;
            else if (content[i] === '}') depth--;
            i++;
        }
        return content.substring(startIndex, i - 1);
    }

    // Extract flow steps from reconcile function
    extractFlowSteps(body) {
        const steps = [];
        
        // Common patterns in reconcile functions
        const patterns = [
            { pattern: /\.\s*Get\s*\(ctx\b[^)]*\)/g, type: 'resource_fetch' },
            { pattern: /\.\s*Create\s*\(ctx\b[^)]*\)/g, type: 'resource_create' },
            { pattern: /\.\s*Update\s*\(ctx\b[^)]*\)/g, type: 'resource_update' },
            { pattern: /\.\s*Delete\s*\(ctx\b[^)]*\)/g, type: 'resource_delete' },
            { pattern: /\.\s*Patch\s*\(ctx\b[^)]*\)/g, type: 'resource_patch' },
            { pattern: /\.Set\s*\(condition\.\w+/g, type: 'condition_set' },
            { pattern: /ctrl\.Result\{[^}]*\}/g, type: 'result_return' },
            { pattern: /controllerutil\.\w+Finalizer/g, type: 'finalizer' },
            { pattern: /helper\.GetConfigMapAndHashWithName/g, type: 'config_map' },
            { pattern: /condition\.CreateList/g, type: 'condition_init' }
        ];
        
        patterns.forEach(({ pattern, type }) => {
            let match;
            while ((match = pattern.exec(body)) !== null) {
                steps.push({
                    type: type,
                    code: match[0],
                    line: this.getLineNumber(body, match.index)
                });
            }
        });
        
        return steps.sort((a, b) => a.line - b.line);
    }

    // Extract error handling patterns
    extractErrorHandling(body) {
        const errorPatterns = [];
        const patterns = [
            /if\s+err\s*!=\s*nil\s*{[^}]*}/g,
            /return\s+[^,]*,\s*err/g,
            /ctrl\.Result\{.*\},\s*err/g
        ];
        
        patterns.forEach(pattern => {
            let match;
            while ((match = pattern.exec(body)) !== null) {
                errorPatterns.push({
                    code: match[0],
                    line: this.getLineNumber(body, match.index)
                });
            }
        });
        
        return errorPatterns;
    }

    // Extract return statements
    extractReturns(body) {
        const returns = [];
        const returnPattern = /return\s+([^;]+)/g;
        
        let match;
        while ((match = returnPattern.exec(body)) !== null) {
            returns.push({
                code: match[0],
                value: match[1].trim(),
                line: this.getLineNumber(body, match.index)
            });
        }
        
        return returns;
    }

    // Find CRD definitions
    findCRDs(dir) {
        const crds = [];
        
        try {
            this.walkDirectory(dir, (filePath) => {
                if (filePath.endsWith('.yaml') || filePath.endsWith('.yml')) {
                    const content = fs.readFileSync(filePath, 'utf8');
                    if (content.includes('kind: CustomResourceDefinition')) {
                        crds.push({
                            file: path.relative(process.cwd(), filePath),
                            content: this.extractCRDInfo(content)
                        });
                    }
                }
            });
        } catch (error) {
            console.error('Error finding CRDs:', error.message);
        }

        return crds;
    }

    // Extract CRD information
    extractCRDInfo(content) {
        const nameMatch = content.match(/name:\s+([^\n]+)/);
        const groupMatch = content.match(/group:\s+([^\n]+)/);
        const kindMatch = content.match(/kind:\s+([^\n]+)/);
        
        return {
            name: nameMatch ? nameMatch[1].trim() : null,
            group: groupMatch ? groupMatch[1].trim() : null,
            kind: kindMatch ? kindMatch[1].trim() : null
        };
    }

    // Find webhook configurations
    findWebhooks(dir) {
        const webhooks = [];
        
        try {
            this.walkDirectory(dir, (filePath) => {
                if (filePath.endsWith('.go')) {
                    const content = fs.readFileSync(filePath, 'utf8');
                    const webhook = this.parseWebhookFile(filePath, content);
                    if (webhook) {
                        webhooks.push(webhook);
                    }
                }
            });
        } catch (error) {
            console.error('Error finding webhooks:', error.message);
        }

        return webhooks;
    }

    // Parse webhook file
    parseWebhookFile(filePath, content) {
        const webhookPatterns = [
            /func\s+\([^)]+\)\s+ValidateCreate\s*\([^)]*\)/g,
            /func\s+\([^)]+\)\s+ValidateUpdate\s*\([^)]*\)/g,
            /func\s+\([^)]+\)\s+ValidateDelete\s*\([^)]*\)/g,
            /func\s+\([^)]+\)\s+Default\s*\([^)]*\)/g
        ];
        
        const webhooks = [];
        
        webhookPatterns.forEach(pattern => {
            let match;
            while ((match = pattern.exec(content)) !== null) {
                webhooks.push({
                    function: match[0],
                    line: this.getLineNumber(content, match.index)
                });
            }
        });
        
        if (webhooks.length > 0) {
            return {
                file: path.relative(process.cwd(), filePath),
                webhooks: webhooks
            };
        }
        
        return null;
    }

    // Find main function
    findMainFunction(dir) {
        try {
            const candidates = [
                path.join(dir, 'main.go'),
                path.join(dir, 'cmd', 'main.go')
            ];
            for (const mainPath of candidates) {
                if (fs.existsSync(mainPath)) {
                    const content = fs.readFileSync(mainPath, 'utf8');
                    return this.parseMainFunction(content);
                }
            }
        } catch (error) {
            console.error('Error finding main function:', error.message);
        }
        
        return null;
    }

    // Parse main function
    parseMainFunction(content) {
        const setupPattern = /mgr\.Add\([^)]+\)/g;
        const controllerPattern = /\.SetupWithManager\([^)]+\)/g;
        
        const setup = [];
        let match;
        
        while ((match = setupPattern.exec(content)) !== null) {
            setup.push({
                code: match[0],
                line: this.getLineNumber(content, match.index)
            });
        }
        
        while ((match = controllerPattern.exec(content)) !== null) {
            setup.push({
                code: match[0],
                line: this.getLineNumber(content, match.index)
            });
        }
        
        return {
            setup: setup.sort((a, b) => a.line - b.line),
            imports: this.extractImports(content)
        };
    }

    // Utility functions
    walkDirectory(dir, callback) {
        if (!fs.existsSync(dir)) return;
        
        const files = fs.readdirSync(dir);
        files.forEach(file => {
            const filePath = path.join(dir, file);
            const stat = fs.statSync(filePath);
            
            if (stat.isDirectory() && !file.startsWith('.') && file !== 'vendor') {
                this.walkDirectory(filePath, callback);
            } else if (stat.isFile()) {
                callback(filePath);
            }
        });
    }

    extractImports(content) {
        const importPattern = /import\s*\(\s*([^)]*)\s*\)/s;
        const match = importPattern.exec(content);
        if (match) {
            return match[1].split('\n').map(line => line.trim()).filter(line => line);
        }
        return [];
    }

    extractStructs(content) {
        const structPattern = /type\s+(\w+)\s+struct\s*{([^}]*)}/g;
        const structs = [];
        
        let match;
        while ((match = structPattern.exec(content)) !== null) {
            structs.push({
                name: match[1],
                fields: match[2].trim(),
                line: this.getLineNumber(content, match.index)
            });
        }
        
        return structs;
    }

    getLineNumber(content, index) {
        return content.substring(0, index).split('\n').length;
    }
}

// CLI interface
if (require.main === module) {
    const args = process.argv.slice(2);
    const dir = args[0] || process.cwd();
    
    if (args.includes('--help') || args.includes('-h')) {
        console.log('Usage: code-parser.js [directory]');
        console.log('  directory: Path to operator code (default: current directory)');
        process.exit(0);
    }
    
    const parser = new OperatorFlowParser();
    parser.parseOperatorDirectory(dir);
}

module.exports = OperatorFlowParser;

File: lib/code-parser.py
#!/usr/bin/env python3

"""
Code Flow Parser for openstack-k8s-operators Operators
Analyzes Go operator code to extract flow patterns
"""

import json
import re
import sys
import os


class OperatorFlowParser:
    def __init__(self):
        self.controllers = []
        self.reconcile_functions = []
        self.custom_resources = []
        self.webhooks = []

    def parse_operator_directory(self, dir_path):
        """Parse operator directory structure"""
        try:
            result = {
                'controllers': self.find_controllers(dir_path),
                'reconcilers': self.find_reconcilers(dir_path),
                'crds': self.find_crds(dir_path),
                'webhooks': self.find_webhooks(dir_path),
                'main': self.find_main_function(dir_path)
            }

            print(json.dumps(result, indent=2))
            return result
        except Exception as e:
            print(f'Error parsing operator: {e}', file=sys.stderr)
            return None

    def find_controllers(self, dir_path):
        """Find controller files"""
        controllers = []
        controller_pattern = re.compile(r'controller|reconciler', re.IGNORECASE)

        try:
            for file_path in self.walk_directory(dir_path):
                if file_path.endswith('.go') and controller_pattern.search(file_path):
                    try:
                        with open(file_path, 'r', encoding='utf-8') as f:
                            content = f.read()
                        controller = self.parse_controller_file(file_path, content)
                        if controller:
                            controllers.append(controller)
                    except Exception as e:
                        print(f'Error reading {file_path}: {e}', file=sys.stderr)
        except Exception as e:
            print(f'Error finding controllers: {e}', file=sys.stderr)

        return controllers

    def parse_controller_file(self, file_path, content):
        """Parse individual controller file"""
        reconcile_pattern = re.compile(
            r'func\s+\(.*?\)\s+Reconcile\s*\([^)]*\)\s+\([^)]*\)',
            re.DOTALL
        )
        setup_pattern = re.compile(
            r'func\s+\([^)]+\)\s+SetupWithManager\s*\([^)]*\)',
            re.DOTALL
        )

        reconcile_functions = []
        setup_functions = []

        # Find Reconcile functions
        for match in reconcile_pattern.finditer(content):
            reconcile_functions.append({
                'signature': match.group(0),
                'line': self.get_line_number(content, match.start())
            })

        # Find SetupWithManager functions
        for match in setup_pattern.finditer(content):
            setup_functions.append({
                'signature': match.group(0),
                'line': self.get_line_number(content, match.start())
            })

        if reconcile_functions or setup_functions:
            return {
                'file': os.path.relpath(file_path, os.getcwd()),
                'reconcile': reconcile_functions,
                'setup': setup_functions,
                'imports': self.extract_imports(content),
                'structs': self.extract_structs(content)
            }

        return None

    def find_reconcilers(self, dir_path):
        """Find reconciler functions and their flow"""
        reconcilers = []

        try:
            for file_path in self.walk_directory(dir_path):
                if file_path.endswith('.go'):
                    try:
                        with open(file_path, 'r', encoding='utf-8') as f:
                            content = f.read()
                        flows = self.parse_reconcile_flow(file_path, content)
                        if flows:
                            reconcilers.append({
                                'file': os.path.relpath(file_path, os.getcwd()),
                                'flows': flows
                            })
                    except Exception as e:
                        print(f'Error reading {file_path}: {e}', file=sys.stderr)
        except Exception as e:
            print(f'Error finding reconcilers: {e}', file=sys.stderr)

        return reconcilers

    def parse_reconcile_flow(self, file_path, content):
        """Parse reconcile function flow"""
        flows = []
        signature_pattern = re.compile(
            r'func\s+\([^)]+\)\s+Reconcile\s*\([^)]*\)\s+\([^)]*\)\s*{',
            re.DOTALL
        )

        for match in signature_pattern.finditer(content):
            body_start = match.end()
            body = self.extract_function_body(content, body_start)
            flow = {
                'function': match.group(0).rstrip().rstrip('{').strip() + ' {...}',
                'line': self.get_line_number(content, match.start()),
                'steps': self.extract_flow_steps(body),
                'errorHandling': self.extract_error_handling(body),
                'returns': self.extract_returns(body)
            }
            flows.append(flow)

        return flows

    def extract_function_body(self, content, start_index):
        """Extract function body using brace counting"""
        depth = 1
        i = start_index
        while i < len(content) and depth > 0:
            if content[i] == '{':
                depth += 1
            elif content[i] == '}':
                depth -= 1
            i += 1
        return content[start_index:i - 1]

    def extract_flow_steps(self, body):
        """Extract flow steps from reconcile function"""
        steps = []

        # Common patterns in reconcile functions
        patterns = [
            (re.compile(r'\.\s*Get\s*\(ctx\b[^)]*\)'), 'resource_fetch'),
            (re.compile(r'\.\s*Create\s*\(ctx\b[^)]*\)'), 'resource_create'),
            (re.compile(r'\.\s*Update\s*\(ctx\b[^)]*\)'), 'resource_update'),
            (re.compile(r'\.\s*Delete\s*\(ctx\b[^)]*\)'), 'resource_delete'),
            (re.compile(r'\.\s*Patch\s*\(ctx\b[^)]*\)'), 'resource_patch'),
            (re.compile(r'\.Set\s*\(condition\.\w+'), 'condition_set'),
            (re.compile(r'ctrl\.Result\{[^}]*\}'), 'result_return'),
            (re.compile(r'controllerutil\.\w+Finalizer'), 'finalizer'),
            (re.compile(r'helper\.GetConfigMapAndHashWithName'), 'config_map'),
            (re.compile(r'condition\.CreateList'), 'condition_init')
        ]

        for pattern, step_type in patterns:
            for match in pattern.finditer(body):
                steps.append({
                    'type': step_type,
                    'code': match.group(0),
                    'line': self.get_line_number(body, match.start())
                })

        return sorted(steps, key=lambda x: x['line'])

    def extract_error_handling(self, body):
        """Extract error handling patterns"""
        error_patterns = []
        patterns = [
            re.compile(r'if\s+err\s*!=\s*nil\s*{[^}]*}'),
            re.compile(r'return\s+[^,]*,\s*err'),
            re.compile(r'ctrl\.Result\{.*\},\s*err')
        ]

        for pattern in patterns:
            for match in pattern.finditer(body):
                error_patterns.append({
                    'code': match.group(0),
                    'line': self.get_line_number(body, match.start())
                })

        return error_patterns

    def extract_returns(self, body):
        """Extract return statements"""
        returns = []
        return_pattern = re.compile(r'return\s+([^;]+)')

        for match in return_pattern.finditer(body):
            returns.append({
                'code': match.group(0),
                'value': match.group(1).strip(),
                'line': self.get_line_number(body, match.start())
            })

        return returns

    def find_crds(self, dir_path):
        """Find CRD definitions"""
        crds = []

        try:
            for file_path in self.walk_directory(dir_path):
                if file_path.endswith('.yaml') or file_path.endswith('.yml'):
                    try:
                        with open(file_path, 'r', encoding='utf-8') as f:
                            content = f.read()
                        if 'kind: CustomResourceDefinition' in content:
                            crds.append({
                                'file': os.path.relpath(file_path, os.getcwd()),
                                'content': self.extract_crd_info(content)
                            })
                    except Exception as e:
                        print(f'Error reading {file_path}: {e}', file=sys.stderr)
        except Exception as e:
            print(f'Error finding CRDs: {e}', file=sys.stderr)

        return crds

    def extract_crd_info(self, content):
        """Extract CRD information"""
        name_match = re.search(r'name:\s+([^\n]+)', content)
        group_match = re.search(r'group:\s+([^\n]+)', content)
        kind_match = re.search(r'kind:\s+([^\n]+)', content)

        return {
            'name': name_match.group(1).strip() if name_match else None,
            'group': group_match.group(1).strip() if group_match else None,
            'kind': kind_match.group(1).strip() if kind_match else None
        }

    def find_webhooks(self, dir_path):
        """Find webhook configurations"""
        webhooks = []

        try:
            for file_path in self.walk_directory(dir_path):
                if file_path.endswith('.go'):
                    try:
                        with open(file_path, 'r', encoding='utf-8') as f:
                            content = f.read()
                        webhook = self.parse_webhook_file(file_path, content)
                        if webhook:
                            webhooks.append(webhook)
                    except Exception as e:
                        print(f'Error reading {file_path}: {e}', file=sys.stderr)
        except Exception as e:
            print(f'Error finding webhooks: {e}', file=sys.stderr)

        return webhooks

    def parse_webhook_file(self, file_path, content):
        """Parse webhook file"""
        webhook_patterns = [
            re.compile(r'func\s+\([^)]+\)\s+ValidateCreate\s*\([^)]*\)'),
            re.compile(r'func\s+\([^)]+\)\s+ValidateUpdate\s*\([^)]*\)'),
            re.compile(r'func\s+\([^)]+\)\s+ValidateDelete\s*\([^)]*\)'),
            re.compile(r'func\s+\([^)]+\)\s+Default\s*\([^)]*\)')
        ]

        webhooks = []

        for pattern in webhook_patterns:
            for match in pattern.finditer(content):
                webhooks.append({
                    'function': match.group(0),
                    'line': self.get_line_number(content, match.start())
                })

        if webhooks:
            return {
                'file': os.path.relpath(file_path, os.getcwd()),
                'webhooks': webhooks
            }

        return None

    def find_main_function(self, dir_path):
        """Find main function"""
        try:
            candidates = [
                os.path.join(dir_path, 'main.go'),
                os.path.join(dir_path, 'cmd', 'main.go')
            ]
            for main_path in candidates:
                if os.path.exists(main_path):
                    with open(main_path, 'r', encoding='utf-8') as f:
                        content = f.read()
                    return self.parse_main_function(content)
        except Exception as e:
            print(f'Error finding main function: {e}', file=sys.stderr)

        return None

    def parse_main_function(self, content):
        """Parse main function"""
        setup_pattern = re.compile(r'mgr\.Add\([^)]+\)')
        controller_pattern = re.compile(r'\.SetupWithManager\([^)]+\)')

        setup = []

        for match in setup_pattern.finditer(content):
            setup.append({
                'code': match.group(0),
                'line': self.get_line_number(content, match.start())
            })

        for match in controller_pattern.finditer(content):
            setup.append({
                'code': match.group(0),
                'line': self.get_line_number(content, match.start())
            })

        return {
            'setup': sorted(setup, key=lambda x: x['line']),
            'imports': self.extract_imports(content)
        }

    # Utility functions
    def walk_directory(self, dir_path):
        """Walk directory and yield file paths, skipping .git, vendor, and dot-directories"""
        if not os.path.exists(dir_path):
            return

        for root, dirs, files in os.walk(dir_path):
            # Skip .git, vendor, and dot-directories
            dirs[:] = [d for d in dirs if not d.startswith('.') and d != 'vendor']

            for file in files:
                yield os.path.join(root, file)

    def extract_imports(self, content):
        """Extract imports from Go file"""
        import_pattern = re.compile(r'import\s*\(\s*([^)]*)\s*\)', re.DOTALL)
        match = import_pattern.search(content)
        if match:
            imports = [line.strip() for line in match.group(1).split('\n')]
            return [imp for imp in imports if imp]
        return []

    def extract_structs(self, content):
        """Extract struct definitions"""
        struct_pattern = re.compile(r'type\s+(\w+)\s+struct\s*{([^}]*)}')
        structs = []

        for match in struct_pattern.finditer(content):
            structs.append({
                'name': match.group(1),
                'fields': match.group(2).strip(),
                'line': self.get_line_number(content, match.start())
            })

        return structs

    def get_line_number(self, content, index):
        """Get line number from string index"""
        return content[:index].count('\n') + 1


# CLI interface
if __name__ == "__main__":
    args = sys.argv[1:]
    dir_path = args[0] if args and not args[0].startswith('-') else os.getcwd()

    if '--help' in args or '-h' in args:
        print('Usage: code-parser.py [directory]')
        print('  directory: Path to operator code (default: current directory)')
        sys.exit(0)

    parser = OperatorFlowParser()
    parser.parse_operator_directory(dir_path)
